File: app/src/main/java/com/example/firebasestudy/MainActivity.java
package com.example.firebasestudy;

import android.os.Bundle;
import android.util.Log;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.ui.AppBarConfiguration;
import androidx.navigation.ui.NavigationUI;

import com.example.firebasestudy.databinding.ActivityMainBinding;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.iid.FirebaseInstanceId;
import com.google.firebase.messaging.FirebaseMessaging;

public class MainActivity extends AppCompatActivity {
    AppBarConfiguration appBarConfiguration ;
    ActivityMainBinding binding;
    private FirebaseAuth mAuth;
    private FirebaseDatabase mDatabase;
    private DatabaseReference mRefrence;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        initFireBase();
        saveToken();
        subscribeTotopic();
        BottomNavigationView navView = binding.navView;


        // Passing each menu ID as a set of Ids because each
        // menu should be considered as top level destinations.
        appBarConfiguration = new AppBarConfiguration.Builder(
                R.id.navigation_home, R.id.navigation_dashboard)
                .build();
        NavController navController = Navigation.findNavController(this, R.id.nav_host_fragment);
        NavigationUI.setupActionBarWithNavController(this, navController, appBarConfiguration);
        NavigationUI.setupWithNavController(navView, navController);
        binding.fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                navController.navigate(R.id.addArticleFragment);
            }
        });

    }

    @Override
    public boolean onSupportNavigateUp() {
        NavController navController = Navigation.findNavController(this, R.id.nav_host_fragment);
        return NavigationUI.navigateUp(navController, appBarConfiguration)
                || super.onSupportNavigateUp();

    }

    private void initFireBase() {
        mAuth = FirebaseAuth.getInstance();
        //get la base de donné
        mDatabase = FirebaseDatabase.getInstance();
        // le curseur
        mRefrence = mDatabase.getReference("user");

    }


    private void saveToken() {
        FirebaseInstanceId.getInstance().getInstanceId().addOnCompleteListener(task -> {
            if (!task.isSuccessful()) {
                Log.e("Main", "getInstanceId failed", task.getException());
                return;
            } else {
                String token = task.getResult().getToken();
                mRefrence.child(mAuth.getUid()).child("token").setValue(token).addOnCompleteListener(new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {
                        if (task.isSuccessful()) {
                            Log.e("token", "token updated :" + token);
                        }
                    }
                });


            }

        });

    }

    private void subscribeTotopic() {

        FirebaseMessaging.getInstance().subscribeToTopic("article").addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if (!task.isSuccessful()) {
                    Log.e("erro", "failed ti subscribe");
                } else {
                    Log.e("topic", "SUcces");
                }
            }
        });


    }



}

File: app/src/main/java/com/example/firebasestudy/MainActivity.kt
package com.example.firebasestudy

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.navigation.Navigation
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.NavigationUI
import com.example.firebasestudy.databinding.ActivityMainBinding
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.iid.FirebaseInstanceId
import com.google.firebase.iid.InstanceIdResult
import com.google.firebase.messaging.FirebaseMessaging

class MainActivity : AppCompatActivity() {
    var appBarConfiguration: AppBarConfiguration? = null
    private lateinit var binding: ActivityMainBinding
    private var mAuth: FirebaseAuth? = null
    private var mDatabase: FirebaseDatabase? = null
    private var mRefrence: DatabaseReference? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        initFireBase()
        saveToken()
        subscribeTotopic()
        val navView = binding.navView


        // Passing each menu ID as a set of Ids because each
        // menu should be considered as top level destinations.
        appBarConfiguration = AppBarConfiguration.Builder(
                R.id.navigation_home, R.id.navigation_dashboard)
                .build()
        val navController = Navigation.findNavController(this, R.id.nav_host_fragment)
        NavigationUI.setupActionBarWithNavController(this, navController, appBarConfiguration!!)
        NavigationUI.setupWithNavController(navView, navController)
        binding.fab.setOnClickListener { navController.navigate(R.id.addArticleFragment) }
    }

    override fun onSupportNavigateUp(): Boolean {
        val navController = Navigation.findNavController(this, R.id.nav_host_fragment)
        return (NavigationUI.navigateUp(navController, this.appBarConfiguration!!)
                || super.onSupportNavigateUp())
    }

    private fun initFireBase() {
        mAuth = FirebaseAuth.getInstance()
        //get la base de donné
        mDatabase = FirebaseDatabase.getInstance()
        // le curseur
        mRefrence = mDatabase?.getReference("user")
    }

    private fun saveToken() {
        FirebaseInstanceId.getInstance().instanceId.addOnCompleteListener { task: Task<InstanceIdResult?>? ->
            if (!task?.isSuccessful!!) {
                Log.e("Main", "getInstanceId failed", task.exception)
                return@addOnCompleteListener
            } else {
                val token: String = task.result?.token!!
                mRefrence?.child(mAuth?.uid!!)?.child("token")?.setValue(token)?.addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        Log.e("token", "token updated :$token")
                    }
                }
            }
        }
    }

    private fun subscribeTotopic() {
        FirebaseMessaging.getInstance().subscribeToTopic("article").addOnCompleteListener { task ->
            if (!task.isSuccessful) {
                Log.e("erro", "failed ti subscribe")
            } else {
                Log.e("topic", "SUcces")
            }
        }
    }
}